fv3gfs/util/halo_updater.py

```diff
@@ -87,33 +87,29 @@ def from_scalar_specifications(
 
         timer = optional_timer if optional_timer is not None else NullTimer()
 
-        exchange_descriptors = defaultdict(list)
+        transformers: Dict[int, HaloExchangeSpec] = {}
+
         for boundary in boundaries:
             for specification in specifications:
-                exchange_descriptors[boundary.to_rank].append(
-                    HaloExchangeData(
+                transformers[boundary.to_rank] = HaloDataTransformer.get(
+                    numpy_like_module, 
+                    HaloExchangeSpec(
                         specification,
                         boundary.send_slice(specification),
                         boundary.n_clockwise_rotations,
                         boundary.recv_slice(specification),
                     )
                 )
 
-        transformers = {}
-        for rank, exchange_descriptor in exchange_descriptors.items():
-            transformers[rank] = HaloDataTransformer.get(
-                numpy_like_module, exchange_descriptor
-            )
-
         return cls(comm, tag, transformers, timer)
```

Comment:
Merged two for loops here

Reply:
You can't merge those loops because you can have multiple specification `QuantityHaloSpec` passed as an argument that you want to bundle in a single `HaloDataTransformer`, that is the point of this technique actually.

So you have to first sort the `QuantityHaloSpec` per target rank than allocate the Transformer. Previously I had the API to do it sequentially using `queue_scalar_spec` but we decided to move away from that for a single `get` with a `Sequence`

Reply:
Reverted and added comments for comprehension